src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs

```diff
@@ -51,70 +50,172 @@ internal sealed class ServerMonitor : IServerMonitor
+            }
+        }
+
+        // private methods
+        private CommandWireProtocol<BsonDocument> InitializeIsMasterProtocol(IConnection connection)
+        {
+            BsonDocument isMasterCommand;
+            var commandResponseHandling = CommandResponseHandling.Return;
+            if (connection.Description.IsMasterResult.TopologyVersion != null)
             {
-                heartbeatDelay.RequestHeartbeat();
+                connection.SetReadTimeout(_serverMonitorSettings.ConnectTimeout + _serverMonitorSettings.HeartbeatInterval);
+                commandResponseHandling = CommandResponseHandling.ExhaustAllowed;
+
+                var veryLargeHeartbeatInterval = TimeSpan.FromDays(1); // the server doesn't support Infinite value, so we set just a big enough value
+                var maxAwaitTime = _serverMonitorSettings.HeartbeatInterval == Timeout.InfiniteTimeSpan ? veryLargeHeartbeatInterval : _serverMonitorSettings.HeartbeatInterval;
+                isMasterCommand = IsMasterHelper.CreateCommand(connection.Description.IsMasterResult.TopologyVersion, maxAwaitTime);
+            }
+            else
+            {
+                isMasterCommand = IsMasterHelper.CreateCommand();
+            }
+
+            return IsMasterHelper.CreateProtocol(isMasterCommand, commandResponseHandling);
+        }
+
+        private async Task<IConnection> InitializeConnectionAsync(CancellationToken cancellationToken) // called setUpConnection in spec
+        {
+            var connection = _connectionFactory.CreateConnection(_serverId, _endPoint);
+
+            var stopwatch = Stopwatch.StartNew();
+            try
+            {
+                // if we are cancelling, it's because the server has
+                // been shut down and we really don't need to wait.
+                await connection.OpenAsync(cancellationToken).ConfigureAwait(false);
             }
+            catch
+            {
+                // dispose it here because the _connection is not initialized yet
+                try { connection.Dispose(); } catch { }
+                throw;
+            }
+            stopwatch.Stop();
+
+            _roundTripTimeMonitor.AddSample(stopwatch.Elapsed);
+            return connection;
         }
 
         private async Task MonitorServerAsync()
         {
-            var metronome = new Metronome(_heartbeatInterval);
-            var heartbeatCancellationToken = _cancellationTokenSource.Token;
-            while (!heartbeatCancellationToken.IsCancellationRequested)
+            await Task.Yield(); // return control immediately
+
+            var metronome = new Metronome(_serverMonitorSettings.HeartbeatInterval);
+            var monitorCancellationToken = _monitorCancellationTokenSource.Token;
+
+            while (!monitorCancellationToken.IsCancellationRequested)
             {
                 try
                 {
+                    CancellationToken cachedCurrentCheckCancellationToken;
```

Comment:
If you agree to rename `_currentCheckCancellationTokenSource` to `_heartbeatCancellationTokenSource` then this variable can be renamed to `cachedHeartbeatCancellationToken` (or `currentHeartbeatCancellationToken`).

Reply:
done